### PR TITLE
[RLlib] DreamerV3 torch: Fix cont. actions learning.

### PR DESCRIPTION
**Pull request by sven1977**
DreamerV3 torch: Fix cont. actions learning.
* Fixed forward pass through world-model for actor loss: This particular forward through the world model must NOT trace any gradients as to not break torch DDP for multiple-GPUs. In torch DDP, more than one pass through the same gradients is not allowed. This is not a problem for discrete actions, b/c there, the actor loss does NOT depend on the world-model results.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/dreamerv3/torch/models/dreamer_model.py [hunk 1]
```diff
@@ -248,10 +252,9 @@ def dream_trajectory(self, start_states, start_is_terminated, timesteps_H, gamma
         a_dreamed_dist_params_H_B = torch.stack(a_dreamed_dist_params_t0_to_H, dim=0)
 
         # Compute r using reward predictor.
-        with torch.no_grad():
-            r_dreamed_H_B = inverse_symlog(
-                self.world_model.reward_predictor(h=h_states_HxB, z=z_states_prior_HxB)
-            )
+        r_dreamed_H_B = inverse_symlog(
```
Comment (sven1977) [inline rllib/algorithms/dreamerv3/torch/models/dreamer_model.py:255]: bug fix: `no_grad()` would break the actor loss for cont. actions.

File: rllib/algorithms/dreamerv3/torch/models/dreamer_model.py [hunk 2]
```diff
@@ -267,11 +270,10 @@ def dream_trajectory(self, start_states, start_is_terminated, timesteps_H, gamma
             del results_HxB
 
         # Compute continues using continue predictor.
-        with torch.no_grad():
-            c_dreamed_HxB = self.world_model.continue_predictor(
-                h=h_states_HxB,
-                z=z_states_prior_HxB,
-            )
+        c_dreamed_HxB = self.world_model.continue_predictor(
```
Comment (sven1977) [inline rllib/algorithms/dreamerv3/torch/models/dreamer_model.py:273]: Same bug fix: `no_grad()` would break the actor loss for cont. actions.

File: rllib/tuned_examples/dreamerv3/atari_100k_dreamerv3.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def _env_creator(cfg):
 tune.register_env("env", _env_creator)
 
 default_config = DreamerV3Config()
-lr_multiplier = (args.num_learners or 1) ** 0.5
+lr_multiplier = args.num_learners or 1
```
Comment (sven1977) [inline rllib/tuned_examples/dreamerv3/atari_100k_dreamerv3.py:53]: lr x 4 works better for 4 GPUs than lr x 4**0.5.

But lr tuning remains very important for multi-GPU training.

File: release/release_tests.yaml [hunk 4]
```diff
@@ -2680,6 +2680,37 @@
...
+
+  frequency: nightly
+  team: rllib
+  cluster:
+    byod:
+      type: gpu
+      post_build_script: byod_rllib_dreamerv3_test.sh
+      runtime_env:
+        - RLLIB_TEST_NO_JAX_IMPORT=1
+        - LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/ray/.mujoco/mujoco210/bin
```
Comment (simonsays1980) [inline release/release_tests.yaml:2697]: Why is this extra library link needed here? I thought installing mujoco via gymnasium should already do the trick?

File: rllib/BUILD [hunk 5]
```diff
@@ -839,6 +839,30 @@ py_test(
     ],
 )
 
+# DreamerV3
+# takes too long (up to 20-30min to learn -200 on 1 GPU)
```
Comment (simonsays1980) [inline rllib/BUILD:842]: What does this mean? Should we disable the test then? Or should it only run at certain testing stages?

File: rllib/algorithms/dreamerv3/torch/models/dreamer_model.py [hunk 6]
```diff
@@ -287,22 +289,46 @@ def dream_trajectory(self, start_states, start_is_terminated, timesteps_H, gamma
         # boundary should not be used for critic/actor learning either.
         dream_loss_weights_H_B = torch.cumprod(gamma * c_dreamed_H_B, dim=0) / gamma
 
-        # Compute the value estimates.
-        v, v_symlog_dreamed_logits_HxB = self.critic(
+        # Reactivate world model gradients.
+        for p in self.world_model.parameters():
```
Comment (simonsays1980) [inline rllib/algorithms/dreamerv3/torch/models/dreamer_model.py:293]: I have to say this feels like an unnatural form of doing the backward pass in torch. Deactivating and activating gradient tracking. How does Dafner do it in Jax/tf?

File: rllib/algorithms/dreamerv3/torch/models/dreamer_model.py [hunk 7]
```diff
@@ -287,22 +289,46 @@ def dream_trajectory(self, start_states, start_is_terminated, timesteps_H, gamma
...
-        v_symlog_dreamed_ema_HxB = self.critic(
-            h=h_states_HxB,
-            z=z_states_prior_HxB,
-            return_logits=False,
-            use_ema=True,
-        )
+        # Compute the EMA net outputs w/o any gradients.
+        with torch.no_grad():
+            v_symlog_dreamed_ema_HxB = self.critic(
+                h=h_states_HxB.detach(),
```
Comment (simonsays1980) [inline rllib/algorithms/dreamerv3/torch/models/dreamer_model.py:327]: Dumb question: Do we even have to detach here? These values have been computed without gradient tracking above, haven't they?

File: rllib/env/wrappers/dm_control_wrapper.py [hunk 8]
```diff
@@ -218,3 +218,20 @@ def render(self, mode="rgb_array", height=None, width=None, camera_id=0):
         width = width or self._width
         camera_id = camera_id or self._camera_id
         return self._env.physics.render(height=height, width=width, camera_id=camera_id)
+
+
+@DeveloperAPI
+class ActionClip(gym.ActionWrapper):
```
Comment (simonsays1980) [inline rllib/env/wrappers/dm_control_wrapper.py:224]: Dumb question: we have already a connector that does action normalization and clipping. Why do we need a wrapper?

File: rllib/tuned_examples/dreamerv3/dm_control_suite_vision_dreamerv3.py [hunk 9]
```diff
@@ -30,16 +33,33 @@
 if args.num_envs_per_env_runner is None:
     args.num_envs_per_env_runner = 4 * (args.num_learners or 1)
 
+parts = args.env.split("/")
+assert len(parts) == 3, (
+    "ERROR: DMC env must be formatted as 'DMC/[task]/[domain]', e.g. "
```
Comment (simonsays1980) [inline rllib/tuned_examples/dreamerv3/dm_control_suite_vision_dreamerv3.py:38]: This is very helpful! We should check for other examples that use this or maybe even throw this message in errors with envs

File: rllib/tuned_examples/dreamerv3/pendulum_dreamerv3.py [hunk 10]
```diff
@@ -8,15 +8,56 @@
...
+    )
+    .reporting(
+        metrics_num_episodes_for_smoothing=(args.num_learners or 1),
+        report_images_and_videos=False,
+        report_dream_data=False,
+        report_individual_batch_item_stats=False,
+    )
+    # See Appendix A.
+    .training(
+        model_size="S",
```
Comment (simonsays1980) [inline rllib/tuned_examples/dreamerv3/pendulum_dreamerv3.py:50]: I am still a bit unsure if we should work with these "sizes", or actually with some default model confits that can be loaded directly via imports? Just thinking loudly.

**Review by simonsays1980 — APPROVED** [review]

File: rllib/algorithms/dreamerv3/torch/models/dreamer_model.py (hunk 7 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/torch/models/dreamer_model.py:327]: Yeah, you are probably right. The `no_grad()` context should take care of this already ...

File: rllib/algorithms/dreamerv3/torch/models/dreamer_model.py (hunk 6 above)
Comment (sven1977) [reply rllib/algorithms/dreamerv3/torch/models/dreamer_model.py:293]: You don't have to do this in JAX, nor in tf, because of the different APIs that these have. In Tf, you can easily use the gradient tape to include/exclude certain forward passes from recording gradients, so you can tape the actor, then NOT tape the WM and critic, then use the tape (and the resulting loss on the CRITIC output (value targets)) to compute gradients for only the ACTOR. Torch is much more painful to handle this stuff in this regard. It took me several days, plus extensive help from GPT4, to figure this one out.

File: rllib/BUILD (hunk 5 above)
Comment (sven1977) [reply rllib/BUILD:842]: Yeah, let's leave this one out for now ...
I added a corresponding nightly Pendulum release test, so that cont. actions will be covered.

File: release/release_tests.yaml (hunk 4 above)
Comment (sven1977) [reply release/release_tests.yaml:2697]: True, we don't really need MuJoCo for these tests ...

File: rllib/tuned_examples/dreamerv3/dm_control_suite_vision_dreamerv3.py (hunk 9 above)
Comment (sven1977) [reply rllib/tuned_examples/dreamerv3/dm_control_suite_vision_dreamerv3.py:38]: It's not a generic requirement from DMC, it's merely something I enforce in this particular tuned example, b/c we are later splitting the env-string into its components, so we can create the actual DMCEnv.
